src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DocumentImportsTracker.cs

```diff
@@ -0,0 +1,160 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Razor.ProjectSystem
+{
+    internal class DocumentImportsTracker
+    {
+        private readonly object _lock;
+
+        private IReadOnlyList<DocumentSnapshot> _imports;
+
+        public DocumentImportsTracker()
+        {
+            _lock = new object();
+        }
+
+        public IReadOnlyList<DocumentSnapshot> GetImports(ProjectSnapshot project, DocumentSnapshot document)
+        {
+            if (project == null)
+            {
+                throw new ArgumentNullException(nameof(project));
+            }
+
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            if (_imports == null)
+            {
+                lock (_lock)
+                {
+                    if (_imports == null)
+                    {
+                        _imports = GetImportsCore(project, document);
+                    }
+                }
+            }
+
+            return _imports;
+        }
+
+        private IReadOnlyList<DocumentSnapshot> GetImportsCore(ProjectSnapshot project, DocumentSnapshot document)
+        {
+            var projectEngine = project.GetProjectEngine();
+            var importFeature = projectEngine.ProjectFeatures.OfType<IImportProjectFeature>().FirstOrDefault();
+            var projectItem = projectEngine.FileSystem.GetItem(document.FilePath);
+            var importItems = importFeature?.GetImports(projectItem).Where(i => i.Exists);
+            var imports = new List<DocumentSnapshot>();
+            foreach (var item in importItems)
+            {
+                if (item.PhysicalPath == null)
+                {
+                    // This is a default import.
+                    var defaultImport = new DefaultImportDocumentSnapshot(project, item);
+                    imports.Add(defaultImport);
+                }
+                else
+                {
+                    var import = project.GetDocument(item.PhysicalPath);
```

Comment:
Note that this doesn't correctly deal with linked files today. I don't expect you to fix, it but I want to make sure you understand why that is.